hutch_python/ipython_session_timer.py

```diff
@@ -0,0 +1,108 @@
+"""
+This module modifies an ``ipython`` shell to automatically close if it has been
+idle for a certain number of hours. Each hutch can configure their conf.yml
+file to set a timeout duration. The default duration is 48 hours.
+"""
+
+import time
+from threading import Thread
+
+max_idle_time = 172800.0  # number of seconds in 48 hours
+
+
+def configure_timeout(session_timer):
+    global max_idle_time
+    if isinstance(session_timer, int) and session_timer > 0:
+        max_idle_time = session_timer
+
+
+class IPythonSessionTimer:
+    '''
+    Class tracks the amount of time the current `InteractiveShell` instance (henceforth
+    called 'user session') has been idle and closes the session if more than 48
+    hours have passed.
+
+    Time is in seconds (floating point) since the epoch began. (In UNIX the
+    epoch started on January 1, 1970, 00:00:00 UTC)
+
+    Parameters
+    ----------
+    ipython : ``IPython.terminal.interactiveshell.TerminalInteractiveShell``
+        The active ``ipython`` ``Shell``, perhaps the one returned by
+        ``IPython.get_ipython()``.
+
+    Attributes
+    ----------
+    curr_time: float
+        The current time in seconds.
+
+    max_idle_time: float
+        The maximum number of seconds a user session can be idle (default is
+        172800.0 seconds or 48 hours).
+
+    last_active_time: float
+        The time of the last user activity in this session.
+
+    idle_time: float
+        The amount of time the user session has been idle.
+    '''
+
+    def __init__(self, ipython):
+        global max_idle_time
+        self.curr_time = 0.0
+        self.max_idle_time = max_idle_time
+        self.last_active_time = 0.0
+        self.idle_time = 0.0
+        self.user_active = False
+        self.ip = ipython
+
+        ipython.events.register('pre_run_cell', self._set_user_active)
+        ipython.events.register('post_run_cell', self._set_user_inactive)
+
+    def _set_user_active(self):
+        self.user_active = True
+        self.last_active_time = time.monotonic()
+
+    def _set_user_inactive(self):
+        self.user_active = False
+        self.last_active_time = time.monotonic()
+
+    def _set_idle_time(self):
+        self.curr_time = time.monotonic()
+        self.idle_time = self.curr_time - self.last_active_time
+
+    def _start_session(self):
+        # Check if idle_time has exceeded max_idle_time
+        while (self.idle_time < self.max_idle_time):
+
+            # Check if user is active once every minute
+            while (self.user_active):
+                time.sleep(60)
+                self.idle_time = 0
+
+            time.sleep(self.max_idle_time - self.idle_time)
+            self._set_idle_time()
```

Comment:
Did using `threading.Timer` not work out? I thought that was the cleanest approach.

I'm trying to step through the logic in my head to see if it makes sense to me, since I'm finding it unintuitive.

- At the start, we initialize `self.max_idle_time` to some large configurable, `self.idle_time` to zero, and `self.user_active` to `False`
- Outer while loop evaluates to `True` at first
- Inner while loop evaluates to `False` at first
- We sleep for a long, long time (the configured `self.max_idle_time`)
- While waiting, some number of `_set_user_active` calls are made (which are followed by `_set_user_inactive` calls), which flip the `user_active` boolean to `True` and then to `False` respectively
- Either way, during each of these calls we're updating `self.last_active_time` to the current time
- After the long sleep, we update the idle time to the number of seconds since the last `_set_user_active` or `_set_user_inactive` call
- Now, we have the first chance to exit the loop, which will happen if it's been long enough since `_set_user_active` or `_set_user_inactive`, or if neither has ever been called.
- Suppose we stay in the loop, then we either have `self.user_active` set to `True`, in which case we do a small loop of re-checking for this every minute until the user command ends, successfully preventing us from exiting, or we drop back into the big sleep (which may be less big this time depending on the situation).

Despite confusing me, I think this generally works. There's a small chance of a race condition if a user starts a long command right before the start of the big sleep on line 83, but otherwise this works out reasonably well.

I think I probably would have opted for creating and interrupting timer objects, because it wouldn't have required maintaining state across while loop iterations, it would just have required me to kill a timer at `_set_user_active` and remake it again at `_set_user_inactive`.

Reply:
There was some comments about the timer not working but it lacks some details https://github.com/pcdshub/hutch-python/pull/385#discussion_r1655729172

Reply:
Yep, I did have an implementation where threading.Timer() is launched by `post_run_cell` and canceled by `pre_run_cell`. I couldn't get this implementation to cancel and relaunch Timer() (it looks like threading.Timer() is launched as a thread so once it's canceled a new one has to be spawned). 

I found a stackoverflow solution using a global t = Timer() function to launch and cancel Timer(), but because it gets called inside the IPythonSessionTimer class I tried using self.T = threading.Timer() to substitute for global t. Unfortunately, I couldn't get this to work, so I went back to the first iteration of ipython_session_timer.py and created a combination that works. Unfortunately, the implementation is kind of unintuitive.

Reply:
I see, it's really strange that creating a new timer with self.T wouldn't work.

Reply:
This was my code for the implementation:
```
class IPythonSessionTimer:
    def __init__(self, ipython):
        global max_idle_time
        self.T = Timer(20.0, self._end_session)

        ipython.events.register('pre_run_cell', self.T.cancel)
        ipython.events.register('post_run_cell', self._start_timer)

    def _end_session(self):
        print("This hutch-python session has timed out and automatically closed. Please start a new session.")

        # Close this ipython session
        ip = get_ipython()
        ip.ask_exit()
        ip.pt_app.app.exit()

    def _new_timer(self):
        global max_idle_time
        print("Timer has started.")
        self.T = Timer(max_idle_time, self._end_session)

    def _start_timer(self):
        self.T.cancel()
        self._new_timer()
        self.T.start()

    def _start_session(self):
        self._new_timer()
```

Reply:
That's a real shame because the above snippet is clean and way easier to understand

Reply:
The only thing I see is that the original self.T is never started in the above snippet, only the subsequent timers are started

Reply:
```
def _start_session(self):
        self.T.start()
```
Tried this and no luck :\

Reply:
Yes, but `self._start_session()` is never called in the snippit.

Reply:
Ah, I forgot to include the def load_ipython_extension(ipython) function in the code above. Updated.